Rework find for spreadsheet & worksheet

Kordishal opened the pull request:
This PR reworks both find for spreadsheet & worksheet.

The spreadsheet.find will now simply call worksheet.find for every worksheet present. The result will be returned as a list of lists with an empty list if no match was found.

Find on worksheet will search the datagrid for the pattern according to the config params. Returns an empty list if no match was found.

If the worksheet is linked it will first update values from remote and update cells after replacement.

nithinmurali commented:
Here you have remove `findReplace`, which is a much faster method for linked worksheets?

Kordishal commented:
Yes. You wrote it makes no sense to have two functions?

And findReplace does not return the cells which were found, but I forgot, that I could use the findReplace instead of cell.update() when linked. I will change this.

Kordishal commented:
My latest change does not actually work properly since the local copy is not synchronized.

Which makes me wonder if it makes really sense to have a `findReplaceRequest` in the first place?

I will now have to update all values instead of just the changed values since findReplace does not return the changed values. 

To make this optimized there needs to be an easy mechanism to collect batchUpdate Requests and a quick way to then synchronize the local sheet again once needed.

Kordishal commented:
This way we can make proper use of the `findReplaceRequest` while keeping the return behavior. Wanted to know if this is a better solution?

Ultimately the `findReplaceRequest` will only really be very good if there is a remote only spreadsheet. (e.g. no need to update the local model or when implementing a lazy synchronization.)

Will have to add tests before you merge.

nithinmurali commented:
I think this is the best way we can make use of findReplace by retaining the return consistent.

Kordishal commented:
Updated the test for find (and squashed a bug in the process).

Wanted to add tests for replace, but they currently do not work, because the worksheet id is not correct when creating a new spreadsheet. And when looking at the code it does not look like the value is actually loaded from remote. But I don't want to fix this here in this PR. (And I am not sure how/where)

nithinmurali commented:
> because the worksheet id is not correct when creating a new spreadsheet. And when looking at the code it does not look like the value is actually loaded from remote. But I don't want to fix this here in this PR. (And I am not sure how/where)

can you please elaborate? you can get the worksheet id from `self.worksheet.id` rt?

Kordishal commented:
There was a bug in my code... used worksheet id instead of spreadsheet id to create.

Need to test it tonight.

Kordishal commented:
Tested and pushed! I think this is ready to merge. 

However I found, that it is possible have batchUpdate return spreadsheet data. But this will require a rework of the way batchUpdate is implemented. I will do this at a later point.

nithinmurali commented:
cool, i will merge once i am done with testing.

ashwarsadh commented:
I have tested both Spreadsheet Find and Worksheet Find.
It is reporting correct Column but Wrong Rows Every time.
My Data has many blank rows also and I feel it is related to that, or there is some bug in calculating Row Number.
For Example if Data is in C13 it replaces C9, and if its in D13 then D9 is replaced.
If data is in D20 then D12 was replaced. Kindly look into it.

nithinmurali commented:
@ashwarsadh  Please create an issue referring to the PR as this is already merged.

ashwarsadh commented:
Hi.
I have also created an issue for this.
Please make a quick fix for this bug.
It is easy to recreate
